lib/throttle-stream.js

```diff
@@ -0,0 +1,153 @@
+  if (!new.target) {
+    return new ThrottleStream(opts)
+  }
+
+  Transform.call(this, opts)
+
+  /**
+   * The function to calculate bytes per second to throttle to
+   * @type {Function}
+   * @public
+   * @memberof ThrottleStream
+   * @name bpsFn
+   * @default null
+   */
+  this.bpsFn = typeof opts?.bps === 'undefined'
+    ? function bpsFn (_elapsedTime, _bytes) {
+      return 16384
+    }
+    : typeof opts.bps === 'number'
+      ? function bpsFn (_elapsedTime, _bytes) {
+        return opts.bps
+      }
+      : opts.bps
+
+  /**
+   * The time that the stream started
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name startTime
+   * @default null
+   */
+  this.startTime = null
+
+  /**
+   * The number of bytes that have been sent through the stream
+   * @type {number}
+   * @public
+   * @memberof ThrottleStream
+   * @name bytes
+   * @default 0
+   */
+  this.bytes = 0
+
+  /**
+   * The buffered chunk.
+   * @type {Buffer}
+   * @private
+   * @memberof ThrottleStream
+   * @name _buffer
+   * @default null
+   */
+  this._buffer = null
+
+  this._interval = null
+
+  /**
+   * Number of bytes allowed to be pushed to the next stream
+   * @type {number}
+   * @private
+   * @default 0
+   */
+  this._allowedBytes = 0
+
+  this._callback = null
+}
+
+ThrottleStream.prototype = Object.create(Transform.prototype, {
+  constructor: {
+    value: ThrottleStream,
+    enumerable: false,
+    writable: true,
+    configurable: true
+  }
+})
+
+/**
+ * The internal `_transform` function is called by the `write` function of the
+ * `Transform` stream base class. This function is not intended to be called
+ * directly.
+ * @param {Buffer} chunk
+ * @param {string} encoding
+ * @param {Function} fn
+ */
+ThrottleStream.prototype._transform = function _transform (chunk, encoding, callback) {
+  this.startTime === null && this._init()
+
+  chunk && (this._buffer = chunk)
+  callback && (this._callback = callback)
+
+  if (this._allowedBytes !== 0) {
+    const chunkLength = this._buffer.length
+    if (chunkLength <= this._allowedBytes) {
+      this._allowedBytes -= chunkLength
+      this.push(this._buffer)
+      this._buffer = null
+      this._callback()
+      return
+    }
+
+    const allowedBytes = this._allowedBytes
+    this._allowedBytes = 0
+    this.push(this._buffer.subarray(0, allowedBytes))
+    this._buffer = this._buffer.subarray(allowedBytes)
+  }
+}
+
+ThrottleStream.prototype._init = function _init () {
+  this.startTime = Date.now()
+
+  this._allowedBytes = this.bpsFn(0, 0)
+  this._interval = setInterval(() => {
+    this._allowedBytes = this.bpsFn((Date.now() - this.startTime) / 1000, this.bytes)
+
+    if (
+      this._allowedBytes !== 0 &&
+      this._buffer !== null
+    ) {
+      setImmediate(() => this._transform())
+    }
+  }, 1000)
+  this.once('end', () => {
+    clearInterval(this._interval)
+    this._buffer = null
+    this._callback = null
+  })
+}
+
+module.exports = {
```

Comment:
maybe infamous triplet should be used here?

Reply:
but I dont expose it?

Reply:
wait, so what runtime code is using it?